packages/flutter/lib/src/material/notch_computers.dart

```diff
@@ -0,0 +1,158 @@
+import 'dart:math' as math;
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/rendering.dart';
+
+/// Computes notches in the outline of a shape.
+///
+/// Typically used to compute notches by a [BottomAppBar].
+abstract class NotchComputer {
+  /// Abstract const constructor. This constructor enables subclasses to provide
+  /// const constructors so that they can be used in const expressions.
+  const NotchComputer();
+
+  /// Returns a path for a notch in the outline of a shape.
+  ///
+  /// The path makes a notch in the host shape that can contain the guest shape.
+  ///
+  /// The `host` is the bounding rectangle for the shape into which the notch will
+  /// be applied. The `guest` is the bounding rectangle of the shape for which we
+  /// are creating a notch in the host.
+  ///
+  /// The `start` and `end` arguments are points on the outline of the host shape
+  /// that will be connected by the returned path.
+  ///
+  /// The returned path may pass anywhere, including inside the guest bounds area,
+  /// and may contain multiple subpaths. The returned path ends at `end` and does
+  /// not end with a [Path.close]. The returned [Path] is built under the
+  /// assumption it will be added to an existing path that is at the `start`
+  /// coordinates using [Path.addPath].
+  Path compute(Rect host, Rect guest, Offset start, Offset end);
+}
+
+/// Computes smooth circular notches.
+///
+/// The notches computed by this class are paths that smoothly connect a top
+/// horizontal edge with a circle circumference.
+///
+/// See also: [compute].
+class CircularNotchComputer implements NotchComputer {
+  /// Constructs a [CircularNotchComputer] that can be used for multiple notch
+  /// computations.
+  const CircularNotchComputer({
+    this.notchMargin = 4.0,
+  });
+
+  /// Minimal margin that will be kept from the guest. Assuming that the guest's
+  /// shape is circular.
+  final double notchMargin;
+
+  /// Returns a path for a smooth circular notch in a horizontal top edge of a shape.
+  ///
+  /// The guest's shape is assumed to be a circle bounded by the [guest]
+  /// rectangle.
+  ///
+  /// The returned path makes a notch in the host top edge that can contain the
+  /// guest shape, and keep a [notchMargin] margin between the notch and the
+  /// guest.
+  ///
+  /// The `host` is the bounding rectangle for the shape into which the notch will
+  /// be applied. The `guest` is the bounding rectangle of the shape for which we
+  /// are creating a notch in the host.
+  ///
+  /// The `start` and `end` arguments are points on the outline of the host shape
+  /// that will be connected by the returned path.
+  @override
+  Path compute(Rect host, Rect guest, Offset start, Offset end) {
+    // The guest's shape is a circle bounded by the guest rectangle.
```

Comment:
Perhaps you should add some sanity check assertions here, host and guest rectangles are not empty.

Reply:
Is it possible that a empty host/guest is actually a valid scenario? (e.g maybe during some scale transition?) are you sure we want to crash in this case (even if only in debug mode)?

Reply:
Good point. I agree that it's better to just gracefully handle the exceptional cases (empty guest or host).